bindings_generator/src/special_methods.rs

```diff
@@ -4,113 +4,61 @@ use heck::SnakeCase;
 use proc_macro2::TokenStream;
 use quote::{format_ident, quote};
 
-pub fn generate_reference_ctor(class: &GodotClass) -> TokenStream {
-    let class_name = format_ident!("{}", class.name);
+pub fn generate_ctor(class: &GodotClass) -> TokenStream {
     let method_table = format_ident!("{}MethodTable", class.name);
-    quote! {
-        // Constructor
-        #[inline]
-        pub fn new() -> Self {
-            unsafe {
-                let gd_api = get_api();
-                let ctor = #method_table::get(gd_api).class_constructor.unwrap();
-                let obj = ctor();
-                object::init_ref_count(obj);
-
-                #class_name {
-                    this: obj
-                }
-            }
-        }
-    }
-}
+    let persistent_ref = class.persistent_ref();
 
-pub fn generate_non_reference_ctor(class: &GodotClass) -> TokenStream {
-    let class_name = format_ident!("{}", class.name);
-    let method_table = format_ident!("{}MethodTable", class.name);
-
-    let documentation = format!(
+    let documentation = if class.is_refcounted() {
+        r#"/// Constructor.
+///
+/// This is a reference-counted type. The returned object is automatically managed
+/// by `Ref`"#
+    } else {
         r#"/// Constructor.
 ///
 /// Because this type is not reference counted, the lifetime of the returned object
 /// is *not* automatically managed.
 /// Immediately after creation, the object is owned by the caller, and can be
 /// passed to the engine (in which case the engine will be responsible for
-/// destroying the object) or destroyed manually using `{}::free`."#,
-        class_name
-    );
+/// destroying the object) or destroyed manually using `GdPtr::free`."#
+    };
 
     quote! {
         #[doc=#documentation]
         #[inline]
-        pub fn new() -> Self {
+        pub fn new() -> #persistent_ref {
```

Comment:
I'm mainly concerned that it may be freed by other scripts:

```rust
let node = Node::new();
owner.add_child(node);
owner.call("other_script_function_that_frees_all_children", &[]);
node.get_path(); // Oops!
```

Reply:
Maybe that's not a problem if we assume that `assume_safe` rules are observed everywhere else but that feels like too much.

Reply:
I think it's okay for now, we can see if there will be common patterns and optimise for that later.